### ISSUE TITLE
How to participate/help out

### ISSUE THREAD

**Issue by max-b** [radiowitness/radiowitness.io#20]
Hey @rhodey this is really cool stuff you've put together here! I remember you working on it for a while at sudoroom, but I didn't realize you had gotten so far and put together a functioning platform. Nice work!

@papazoga and I were talking the other day about wanting to plug in/help out with this. Do you have any set tasks that you're thinking of? 

Also - I didn't see any documentation for adding new radios/locations. Perhaps you could point us in some general directions and we could hack away on it and maybe write some documentation as we go along? 

It's really cool - I'm so glad you stuck with it and built this!

**Comment by rhodey** [comment]
Hey @max-b & @papazoga, thanks for the kind words! I'd really love some help with adding a streaming audio feature, something where you could listen to a live feed on a per-city basis. I think that'd be good task for collaboration because it hasn't been started at all so minimal need for coordination :D 

Right now Radio Witness only supports [P25 radios](https://en.wikipedia.org/wiki/Project_25), which is a digital packet-type radio. The stations on the ground do the work of decoding RF into P25 packets and then the packets are batched together and put into an [Amazon Kinesis Stream](https://aws.amazon.com/kinesis/). Before sending the packets to Kinesis they are encoded into a [Cap'n Proto](https://capnproto.org/) buffer and then base64 encoded.

This weekend I'm hoping to switch from Cap'n Proto to protobuf's to improve language support, then you could create a server in whatever language you'd like to:

  1. consume the Kinesis stream
  2. decode & de-batch the "Kinesis records" into P25 packets
  3. synthesize the packets into audio
  4. multiplex the audio to clients

Is this something that'd ya'll would be interesting in working on?

**Comment by max-b** [comment]
Yeah I'd be happy to work on that. I do think I'm a little confused about what you currently have wired up. Maybe you could help me with a few questions?

- What's actually responding to requests at https://radiowitness.io/api ? 

- How would access control to the Kinesis stream work in the scenario you've described?

- Can you elaborate a little bit more on what you're envisioning for the streaming audio feature? Do you feel strongly about the implementation of that #4 step above? What are you thinking that "clients" would be? Are you thinking something along the lines of what [radio reference considers a feed](http://wiki.radioreference.com/index.php/Broadcasting_Feeds)? 


Ok sorry about these pretty basic questions, but I'm just trying to get a handle on things and I'm struggling a little bit to get my bearings here.

**Comment by rhodey** [comment]
cool! yeah no worries there are a lot of repositories and not a lot of documentation, additionally the API-side of `radiowitness.io` is repository private right now, I just added you and @papazoga as collaborators though.

## Architecture
|[SDR (raw samples)](https://github.com/radiowitness/uhd-java)| -> |[chnlzr-server (resampling)](https://github.com/radiowitness/chnlzr-server)| -> |[p25dcodr (decoding)](https://github.com/radiowitness/p25dcodr)| -> |[Kinesis (transport)](https://aws.amazon.com/kinesis/)| -> |[imbeef (audio synthesis)](https://github.com/radiowitness/imbeef)| -> ...

Then `imbeef` pulls all the P25 packets from Kinesis, synthesizes the [IMBE](https://en.wikipedia.org/wiki/Multi-Band_Excitation) vocoder codewords into a WAV audio file, and sends that to Amazon S3. Amazon S3 lets you store small key-value-pair file metadata so additionally `imbeef` tags the WAV file with all available metadata like timestamp, radio ID, talkgroup ID, etc.

There's one more bit of magic I'll try to open source this week, it's called `imbeef-rds-bridge`. `imbeef-rds-bridge` is an [Amazon Lambda](https://aws.amazon.com/lambda/) function that executes every time a WAV file is stored in the Radio Witness S3 bucket. All the function does is read the metadata from the S3 WAV audio file and index it in a Postgres database. This Postgres DB + Amazon S3 is what the `radiowitness.io` API is built upon, you can view the [schema here](https://github.com/radiowitness/schemas/blob/master/v1/tables.sql).

## Kinesis Access Control
Kinesis access control comes down to me creating a new user account, giving you the access & secret keys, and then marking that account as privileged to read from Kinesis. There should be an open source Kinesis client library for pretty much any language, you'll just provide the access & secret keys to some function and then it'll authorize you. If Kinesis did not have a read rate limit I'd just publish read-privileged credentials publicly but right now seems like it's gotta stay kinda locked down.

## Audio Feed
Yeah something like the Radio Reference audio feeds is what comes to mind, but we have these two additional pieces of metadata they don't (radio & talkgroup ID) and so it'd be great if we could propagate those into the feed UX somehow. On radiowitness.io I'm just using this metadata to color the weird audio bubbles right now.

## Next Steps
I worked on switching from Capn' Proto to Google Protobufs this weekend, you can check out the protobuf schema here in the [radiowitness-proto repo](https://github.com/radiowitness/radiowitness-proto/tree/master/src/main/proto). I've pushed these new protobufs into `p25dcodr`, `imbeef`, and `imbeef-rds-bridge` master branches over the weekend but am going to wait till late tonight to try rolling them out into production. Once they're in production I can create Kinesis access keys for you and then using the .proto files you should be able to start streaming P25 packets from Kinesis.

Radio Witness uses a lot of Amazon infrastructure, and yeah it's not the coolest, but it's saving me a hell of a lot of admin time right now. We'll definitely move off of S3 eventually and when that goes so will our use of Lambda, Kinesis is **crazy cheap** and I really love it but maybe we could host something like this ourselves down the line.

**Comment by max-b** [comment]
Ok yay this is making a lot more sense to me now. Thanks for spelling it all out :)

I can totally understand why you went with Amazon services, especially at this stage of dev. I definitely don't blame you for not wanting to spend all that time admin-ing, especially while things are still up in the air. Would be fun eventually to see what sorts of alternatives to these sorts of services exist, but I'm all for the practical.

I think I've got a handle on what my next steps here would be. When you're ready with the IAM user account credentials you can send them over signal. It's been a long time since I've used Java and my understanding of the ecosystem is pretty foggy, so I think I might go with python or Node. There are pretty full featured looking libraries for both, so I'd be covered there. 

It does seem like it would make sense to re-use your P25 code/libraries, especially [p25-common](https://github.com/radiowitness/p25-common)... I need to think a little more about the best way to go about that.

**Comment by rhodey** [comment]
I'd love to get some Node code into the Radio Witness ecosystem, I think it'd be a great fit for this! I should have AIM credentials for you by tomorrow.

p25-common is a good reference but you'll only need about half of it for audio streaming, much of p25-common is for all the work that comes before you can send a decoded P25 packet to Kinesis. I think the [p25-common/protocol package](https://github.com/radiowitness/p25-common/tree/master/src/main/java/org/anhonesteffort/p25/protocol) is most of what's needed here.

Check out the [KinesisP25Consumer::process](https://github.com/radiowitness/imbeef/blob/master/src/main/java/org/anhonesteffort/p25/consumer/KinesisP25Consumer.java#L57) method, this method basically gets called once for every Protocol Buffer message that `imbeef` pulls from Kinesis. `BaseMessage` is a class that gets auto-generated from the [base.proto](https://github.com/radiowitness/radiowitness-proto/blob/master/src/main/proto/base.proto) definition, then it's transformed into an actual P25 Packet / Frame object through [DataUnitFactory::create](https://github.com/radiowitness/imbeef/blob/master/src/main/java/org/anhonesteffort/p25/consumer/DataUnitFactory.java#L40).

**Comment by rhodey** [comment]
I guess the best place to start is just get familiar with [imbeef](https://github.com/radiowitness/imbeef) except when you'd write it to a WAV file and send to S3 you're gonna be streaming it somewhere in some other format.

**Comment by max-b** [comment]
I think one of the real tricks looks like it's going to be IMBE decoding. Apart from writing an entirely new implementation in another language it looks like the two options are the [JMBE library you forked](https://github.com/radiowitness/jmbe) or the [C library](https://github.com/szechyjs/mbelib) that I think it was inspired by. I've been tinkering with node bindings for the C library which I think might end up working out, but I haven't gotten deep enough yet to tell whether or not it will ultimately be the right direction.

**Comment by max-b** [comment]
Hey @rhodey thanks for hooking up some AWS credentials!

I've got some code which can read from the kinesis stream and can decode the protobuf into a BaseMessage.

I've also got some code which can synthesize IMBE to PCM audio (and write it to a wav file as a test).

What I seem to be missing to some extent, is what [bytes](https://github.com/radiowitness/radiowitness-proto/blob/master/src/main/proto/p25.proto#L30) refers to in the `P25DataUnit`. It seems that the length of that byte array is usually 39 or 169, which I can't quite make sense of. I've been trying to parse the imbeef code and the jmbe code to try and understand, but I'm struggling here and could use a bit of help. Thanks!

**Comment by max-b** [comment]
Ok actually I can see the DataUnit class here https://github.com/radiowitness/p25-common/tree/master/src/main/java/org/anhonesteffort/p25/protocol/frame

I'm struggling to parse some of the code - would you be able to give a short high level description for how to go from P25 messages to IMBE data? Is there a good reference somewhere that you'd recommend?

**Comment by rhodey** [comment]
this is my favorite PDF :D took me forever to fine

[p25_training_guide_extended.pdf](https://github.com/radiowitness/radiowitness.io/files/669034/p25_training_guide_extended.pdf)
